### PR TITLE
[QoS] RateLimitExceededException, Mapper and Tests

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**: Throw something better than a generic RuntimeException

**Implementation Description (bullets)**: 
- Wrap RLEEs from Cassandra. But we need to be careful because these get wrapped in an ADDE, unless we take steps to stop that.
- I kept the old unwrap method even though it's not used now, because it will be used by timelock exceptions (I think?)
- I ran a local smoke test and confirmed that I did get a 429 from the AtlasDbEteServer.

**Concerns (what feedback would you like?)**:
- I don't register the RLEE mapper by default, does that make sense?
- Are there any other obvious pitfalls?
- The method name is a... mouthful.

**Where should we start reviewing?**: `Throwables` probably, and the new tests for that

**Priority (whenever / two weeks / yesterday)**: today

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2707)

### CODE REVIEW

File: qos-service-api/src/main/java/com/palantir/atlasdb/qos/ratelimit/RateLimitExceededException.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.qos.ratelimit;
+
+public class RateLimitExceededException extends RuntimeException {
```
Comment (nziebart) [inline qos-service-api/src/main/java/com/palantir/atlasdb/qos/ratelimit/RateLimitExceededException.java]: nit: add some docs here?

File: atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java [hunk 2]
```diff
@@ -79,6 +80,22 @@ public static RuntimeException rewrapAndThrowUncheckedException(String newMessag
         throw createPalantirRuntimeException(newMessage, ex);
     }
 
+    /**
+     * If the provided Throwable is
+     *   a) an ExecutionException or InvocationTargetException, then apply this method on the cause;
+     *   b) a RateLimitExceededException or an AtlasDbDependencyException, then rethrow it;
+     *   c) none of the above, then wrap it in an AtlasDbDependencyException and throw that.
+     */
+    public static RuntimeException unwrapAndThrowRateLimitExceededOrAtlasDbDependencyException(Throwable ex) {
```
Comment (nziebart) [inline atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java]: should we just replace `unwrapAndThrowAtlasDbDependencyException` with this method?

Comment (nziebart) [reply atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java]: is there ever a case where we'd want to use that one and not this one?

File: atlasdb-commons/src/test/java/com/palantir/common/base/ThrowablesTest.java [hunk 3]
```diff
@@ -61,7 +71,46 @@ public void testRewrap() {
...
+    @Test
+    public void unwrapAndThrowRateLimitExceededOrAtlasDbDependencyExceptionThrowsWrappedRateLimitExceededExceptions() {
+        assertThatThrownBy(() -> Throwables.unwrapAndThrowRateLimitExceededOrAtlasDbDependencyException(
+                new ExecutionException(RATE_LIMIT_EXCEEDED_EXCEPTION))).isEqualTo(RATE_LIMIT_EXCEEDED_EXCEPTION);
+        assertThatThrownBy(() -> Throwables.unwrapAndThrowRateLimitExceededOrAtlasDbDependencyException(
+                new InvocationTargetException(RATE_LIMIT_EXCEEDED_EXCEPTION))).isEqualTo(RATE_LIMIT_EXCEEDED_EXCEPTION);
+    }
+
+    @Test
+    public void unwrapAndThrowRateLimitExceededOrAtlasDbDependencyExceptionThrowsWrappedAtlasDbDependencyExceptions() {
```
Comment (hsaraogi) [inline atlasdb-commons/src/test/java/com/palantir/common/base/ThrowablesTest.java]: Great job on adding tests for these, some were beyond the scope of this PR :)

This one test looks repeated, can we delete this.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import javax.ws.rs.ext.ExceptionMapper;
+
+import com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException;
+
+/**
+ * This class maps instances of {@link RateLimitExceededException} to HTTP responses with a 429 status code.
+ * Users may register this exception mapper for a standard way of converting these exceptions to 429 responses, which
+ * are meaningful in that {@link com.palantir.remoting3.jaxrs.JaxRsClient}s are able to back off accordingly.
+ */
+public class RateLimitExceededExceptionMapper implements ExceptionMapper<RateLimitExceededException> {
```
Comment (hsaraogi) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java]: Should we register this by default?

File: atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java]: Yeah, for TimeLock (which isn't QoS serviced I think).

File: qos-service-api/src/main/java/com/palantir/atlasdb/qos/ratelimit/RateLimitExceededException.java (hunk 1 above)
Comment (jeremyk-91) [reply qos-service-api/src/main/java/com/palantir/atlasdb/qos/ratelimit/RateLimitExceededException.java]: Ack, will do after the ongoing superpriority ticket

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java]: Main concern is that this writes exceptions in the remoting2 style, but I suspect some internal users will want to write their exceptions in the remoting3 style.

@nziebart any thoughts?

File: atlasdb-commons/src/test/java/com/palantir/common/base/ThrowablesTest.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-commons/src/test/java/com/palantir/common/base/ThrowablesTest.java]: Actually this test should have used the `ATLASDB_DEPENDENCY_EXCEPTION`, I went too fast and forgot to change it :) good spot!

**Comment by jeremyk-91** [comment]
oh crap, this causes the dependencies to explode 😞 

I might just write a bespoke version of the dependency exception things, rather than adding a dependency from the QoS API to commons

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java (hunk 4 above)
Comment (nziebart) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/http/RateLimitExceededExceptionMapper.java]: i think we've decided it's not atlas' place to modify the API contract of the consumer app, so we shouldn't register by default

**Comment by nziebart** [comment]
My main concern around leaving the old `unwrapAndThrowAtlasDBDependencyException` is that people will use that and it will miss a rate limiting exception. It doesn't hurt to always use `unwrapAndThrowRateLimitExceededOrAtlasDbDependencyException`

**Comment by gsheasby** [comment]
Superceded by #2715.

**Pull request closed without merging**